### PR TITLE
Adding LR scheduling and WD (coupled and decoupled)

### PR DESCRIPTION
**Pull request by zaccharieramzi**
This should solve the issues of regularization and train loss stagnation

### CODE REVIEW

File: solvers/sgd_torch.py [hunk 1]
```diff
@@ -13,16 +13,17 @@ class Solver(TorchSolver):
     # any parameter defined here is accessible as a class attribute
     parameters = {
         'nesterov, momentum': [(False, 0), (True, 0.9)],
-        'lr': [1e-3],
```
Comment (tomMoral) [inline solvers/sgd_torch.py:16]: pour la SGD, je pense qu'utiliser `1e-1` c'est mieux, sinon c'est trop lent non?

Comment (zaccharieramzi) [reply solvers/sgd_torch.py:16]: sure!

File: solvers/sgd_torch.py [hunk 2]
```diff
@@ -14,14 +14,18 @@ class Solver(TorchSolver):
     parameters = {
         'nesterov, momentum': [(False, 0), (True, 0.9)],
         'lr': [1e-1],
+        'coupled_weight_decay': [0.0, 1e-4],
+        'decoupled_weight_decay': [0.0, 1e-4],
```
Comment (tomMoral) [inline solvers/sgd_torch.py:18]: I would simply put `weight_decay` to avoid duplicating the run.
```suggestion
        'weight_decay': [0.0, 1e-4],
```

File: solvers/sgd_torch.py [hunk 3]
```diff
@@ -14,14 +14,18 @@ class Solver(TorchSolver):
...
 
     def set_objective(self, **kwargs):
         super().set_objective(**kwargs)
+        wd = self.coupled_weight_decay or self.decoupled_weight_decay
         self.optimizer_klass = SGD
         self.optimizer_kwargs = dict(
             lr=self.lr,
             momentum=self.momentum,
             nesterov=self.nesterov,
+            weight_decay=wd,
```
Comment (tomMoral) [inline solvers/sgd_torch.py:30]:
```suggestion
        self.optimizer_klass = SGD
        self.optimizer_kwargs = dict(
            lr=self.lr,
            momentum=self.momentum,
            nesterov=self.nesterov,
            weight_decay=self.weight_decay,
```

File: solvers/rmsprop_torch.py [hunk 4]
```diff
@@ -16,6 +16,7 @@ class Solver(TorchSolver):
         'lr': [1e-3],
         'rho': [0.99, 0.9],
         'momentum': [0, 0.9],
+        'coupled_weight_decay': [0.0, 0.02],
```
Comment (tomMoral) [inline solvers/rmsprop_torch.py:19]:
```suggestion
        'weight_decay': [0.0, 0.02],
```

File: solvers/rmsprop_torch.py [hunk 5]
```diff
@@ -26,4 +27,5 @@ def set_objective(self, **kwargs):
             lr=self.lr,
             momentum=self.momentum,
             alpha=self.rho,
+            weight_decay=self.coupled_weight_decay,
```
Comment (tomMoral) [inline solvers/rmsprop_torch.py:30]:
```suggestion
            weight_decay=self.weight_decay,
```

File: utils/tf_solver.py [hunk 6]
```diff
@@ -60,7 +101,31 @@ def get_next(stop_val):
...
+            target_regularizers = [
+                'kernel_regularizer',
+                'bias_regularizer',
+                'beta_regularizer',
+                'gamma_regularizer',
+            ]
+            for layer in self.model.layers:
+                for attr in target_regularizers:
+                    if hasattr(layer, attr):
+                        setattr(layer, attr, regularizer)
```
Comment (tomMoral) [inline utils/tf_solver.py:128]: This seems very custom, why not use the `weight_decay` from the optimizer?

File: utils/tf_solver.py [hunk 7]
```diff
@@ -60,7 +101,31 @@ def get_next(stop_val):
 
     def run(self, callback):
         self.model = self.model_init_fn()
-        self.optimizer = self.optimizer_klass(**self.optimizer_kwargs)
+        self.optimizer = self.optimizer_klass(
+            # this scaling is needed as in TF the weight decay is
+            # not multiplied by the learning rate
+            weight_decay=self.decoupled_weight_decay*self.lr,
+            **self.optimizer_kwargs,
```
Comment (tomMoral) [inline utils/tf_solver.py:108]: For the learning rate, it seems more classical to give `learning_rate=lr_schedule` no?

**Review by tomMoral — COMMENTED** [review]
A few comments here, mostly to make sure I understand correctly.

My main concern is with the implementation for `tf`. I find it very convoluted and I fear that we might  introduce a but. From my readings, it seems we can do something simpler with a single `lr_scheduler`, passing functions to `learning_rate` and `weight_decay` optimizer parameters.
But maybe you already looked into it and it is not doble?

File: utils/tf_solver.py (hunk 7 above)
Comment (tomMoral) [reply utils/tf_solver.py:108]: `lr_schedule` changes the learning rate every iteration and not every epoch.
This is why we need to use the `LearningRateScheduler` API which works with epochs.

File: utils/tf_solver.py (hunk 6 above)
Comment (tomMoral) [reply utils/tf_solver.py:128]: There is no `weight_decay` argument in `Adam`, so we need to use this API to get weight decay.

File: solvers/rmsprop_tf.py [hunk 8]
```diff
@@ -16,6 +16,8 @@ class Solver(TFSolver):
         'lr': [1e-3],
         'rho': [0.99, 0.9],
         'momentum': [0, 0.9],
+        'decoupled_weight_decay': [0.0, 1e-4, 0.02],
+        'coupled_weight_decay': [0.0, 1e-4, 0.02],
```
[the remark this answers is not in the document]
Comment (zaccharieramzi) [reply solvers/rmsprop_tf.py:20]: Let's keep these ones and modify the torch ones.

File: solvers/sgd_torch.py (hunk 2 above)
Comment (zaccharieramzi) [reply solvers/sgd_torch.py:18]: Actually here I wanted to duplicate the run, in order to have an easier time groupping the plots. Indeed since the 2 are equivalent for SGD, it would make sense to compare each one to their application to Adam, wdyt?

File: solvers/rmsprop_torch.py (hunk 4 above)
Comment (zaccharieramzi) [reply solvers/rmsprop_torch.py:19]: I would rather stick with coupled weight decay to clearly mark the difference with the decoupled weight decay, raising awareness that there is a difference between the 2.

File: utils/tf_helper.py [hunk 9]
```diff
@@ -11,3 +14,55 @@ def __init__(self, callback):
...
+        In this case, they mention a per-step schedule, but in order to stick
+        to the PyTorch Lightning way of updating LR and WD, we stick to a
+        per-epoch schedule which is conveniently done with a Callback.
+        """
+        def __init__(self, lr_schedule, wd_schedule, verbose=0):
+            super(LRWDSchedulerCallback, self).__init__(
+                schedule=lr_schedule,
+                verbose=verbose,
+            )
+            self.wd_schedule = wd_schedule
```
Comment (tomMoral) [inline utils/tf_helper.py:34]: Not used no?

File: utils/tf_helper.py [hunk 10]
```diff
@@ -11,3 +14,55 @@ def __init__(self, callback):
...
+            )
+            if self.verbose > 0:
+                io_utils.print_msg(
+                    f'\nEpoch {epoch + 1}: LearningRateScheduler'
+                    ' setting learning '
+                    f'rate to {wd}.')
+
+        def on_epoch_end(self, epoch, logs=None):
+            super().on_epoch_end(epoch, logs)
+            logs['wd'] = backend.get_value(self.model.optimizer.weight_decay)
```
Comment (tomMoral) [inline utils/tf_helper.py:68]: I would remove this as this is not necessary, let's keep the code as small as possible? But if you think it is better to keep it, feel free to ignore this suggestion.
```suggestion
```

**Review by tomMoral — COMMENTED** [review]
Ok I finally got my head around it. I understand what's the point now (it was written in the comment all this time ^^).
Just a question on the `wd_schedule` which seems to be unused.

File: utils/tf_helper.py (hunk 10 above)
Comment (zaccharieramzi) [reply utils/tf_helper.py:68]: I am fine with removing the io, which was faulty anyway given I had not changed the message correctly, but I am in favor of keeping the wd in the logs to inspect it in TensorBoard

File: utils/tf_helper.py (hunk 9 above)
Comment (zaccharieramzi) [reply utils/tf_helper.py:34]: you are right, just changed

Comment (zaccharieramzi) [reply utils/tf_helper.py:34]: (this would have been a pain to debug, glad you caught it)

File: utils/tf_helper.py (hunk 10 above)
Comment (tomMoral) [reply utils/tf_helper.py]: fine for me

**Review by tomMoral — APPROVED** [review]
Ok, LGTM!

Thx @zaccharieramzi :)
